docs/en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_ReactJS_Redux_and_GraphQL.md

```diff
@@ -0,0 +1,666 @@
+        {this.props.items.map(item => (
+          <ItemComponent title={item.title} image={item.image} />
+        ))}
+      </div>
+    );
+  }
+}
+
+export default inject(
+  Gallery, 
+  ['GalleryItem', 'SearchBar'], 
+  (GalleryItem, SearchBar) => ({
+    ItemComponent: GalleryItem,
+    SearchComponent: SearchBar
+  })
+ );
+```
+
+## Using the injector directly within your component
+
+On rare occasions, you may just want direct access to the injector in your component. If
+your dependency requirements are dynamic, for example, you won't be able to explicitly
+declare them in `inject()`. In cases like this, use `withInjector()`. This higher order
+component puts the `Injector` instance in `context`.
+
+```js
+class MyGallery extends React.Component {
+  render () {
+    <div>
+      {this.props.items.map(item => {
+        const Component = this.context.injector.get(item.type);
+        return <Component title={item.title} image={item.image} />
+      })}
+    </div>
+  }
+}
+
+export default withInjector(MyGallery);
+```
+
+## Using Injector to customise forms
+
+Forms in the React layer are built declaratively, using the `FormSchema` API. A component called `FormBuilderLoader` is given a URL to a form schema definition, and it populates itself with fields (both structural and data-containing) and actions to create the UI for the form. Each form is required to have an `identifier` property, which is used to create context for Injector when field components are fetched. This affords developers the opportunity provide very surgical customisations.
+
+### Updating the form schema
+
+Most behavioural and aesthetic customisations will happen via a mutation of the form schema. For this, we'll use the `updater.form.alterSchema()` function.
+
+```js
+Injector.transform(
+	'my-custom-form',
+	(updater) => {
+		update.form.alterSchema(
+			'AssetAdmin.*',
+			(updateSchema) => (form, values) => (
+				updateSchema(
+					form.updateField('Title', {
+						myCustomProp: true
+					})
+					.getState()
+				)
+			)
+		)
+	}
+);
+```
+
+The `alterSchema()` function takes a callback, with an instance of `SchemaStateManager` (`form` in the above example), and a map of the current form values as parameters. `SchemaStateMangaer` allows you to declaratively update the form schema API using several helper methods, including:
+
+* `updateField(fieldName:string, updates:object)`
+* `updateFields({ myFieldName: updates:object })`
+* `mutateField(fieldName:string, callback:function)`
+* `setFieldComponent(fieldName:string, componentName:string)`
+* `setFieldClass(fieldName:string, cssClassName:string, active:boolean)
+* `addFieldClass(fieldName:string, cssClassName:string)`
+* `removeFieldClass(fieldName:string, cssClassName:string)`
+
+<div class="notice" markdown="1">
+It is critical that you end series of mutation calls with `getState()`. The form schema middleware expects an object, and will not handle an instance of `SchemaStateManager`.
```

Comment:
> and will not handle an instance of `SchemaStateManager`

It's a bit confusing to mention a wrong return value by name here. In such a deeply nested callback and argument structure that confused me at least. I'd just leave out the remark